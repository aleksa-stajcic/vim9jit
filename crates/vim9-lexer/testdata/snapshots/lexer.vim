vim9script

var a = 5
var b = a < 5
var c = a > 5
var d = <args>

var num = 1'000
