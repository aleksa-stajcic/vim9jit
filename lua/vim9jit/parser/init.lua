local p = require "vim9jit.patterns"
local g = require "vim9jit.grammar"

local lib = require "vim9jit.lib"

local any_amount = lib.any_amount
local letter = lib.letter
local list_of = lib.list_of
local literal = lib.literal
local one_or_more = lib.one_or_more
local one_or_no = lib.one_or_no
local optional = lib.optional
local range = lib.range
local seq = lib.seq
local set = lib.set
local some = lib.some

-- local whitespace = lib.whitespace

local underscore = literal "_"
local digit = range { "0", "9" }

local any_whitespace = lib.any_whitespace
local any_whitespace_or_eol = lib.any_whitespace_or_eol

local c_seqw = function(sequence)
  local spaces = any_whitespace
  if sequence.eol then
    spaces = any_whitespace_or_eol
  end

  local sequence_with_whitespace = { spaces }
  for i, v in ipairs(sequence) do
    table.insert(sequence_with_whitespace, v)

    -- Don't allow newlines until we say so explicitly
    if i == #sequence then
      table.insert(sequence_with_whitespace, any_whitespace)
    else
      table.insert(sequence_with_whitespace, spaces)
    end
  end
  return p.capture(p.concat(unpack(sequence_with_whitespace)))
end

local EOL = p.end_of_line
local EOL_or_EOF = p.branch(EOL, p.end_of_file)

local group = setmetatable({}, {
  __index = function(_, k)
    return p.group(k)
  end,
})

local make_grammar = function(root)
  local result = {
    root or "vim9script",

    vim9script = seq {
      "vim9script",
      EOL_or_EOF,
      any_amount {
        set {
          seq { group.Assign, EOL_or_EOF },
          -- EOF,
        },
      },

      capture = true,
    },

    Assign = c_seqw {
      group.Variable,
      "=",
      group.Expression,

      eol = false,
    },

    -- VarKeyword    = literal "var",
    -- ConstKeyword  = literal "const",
    -- FinalKeyworld = literal "final",

    -- Var = seq {
    --   set { group.VarKeyword, group.ConstKeyword, group.FinalKeyworld },

    --   capture = true,
    -- },

    -- TypeDefiniton = seq {
    --   ":", group.TypeSyntax
    -- },

    -- TypeSyntax = list_of {
    --   group.Type_SingleDefinition,

    --   separator = "|",
    -- },

    -- Type_SingleDefinition = set {
    --   'asdf'
    -- },

    VariableIdentifier = seq {
      set {
        letter,
        underscore,
      },
      any_amount {
        set { letter, underscore, digit },
      },

      capture = true,
    },

    BufferVariableIdentifier = seq { "b:", group.VariableIdentifier, capture = true },
    GlobalVariableIdentifier = seq { "g:", group.VariableIdentifier, capture = true },
    ScriptVariableIdentifier = seq { "s:", group.VariableIdentifier, capture = true },
    TabVaraibleIdentifier = seq { "t:", group.VariableIdentifier, capture = true },
    VimVariableIdentifier = seq { "v:", group.VariableIdentifier, capture = true },
    WindowVariableIdentifier = seq { "w:", group.VariableIdentifier, capture = true },

    Variable = set {
      group.BufferVariableIdentifier,
      group.GlobalVariableIdentifier,
      group.ScriptVariableIdentifier,
      group.TabVaraibleIdentifier,
      group.VimVariableIdentifier,
      group.WindowVariableIdentifier,

      group.VariableIdentifier,
    },

    --
    -- TODO:
    -- - exponentiation

    --[[
      exp = lpeg.V("term") + lpeg.V("factor") + integer,
      term = node((lpeg.V("factor") + integer) * addsub * lpeg.V("exp")),
      factor = node(integer * muldiv * (lpeg.V("factor") + integer))
    --]]

    Expression = seq {
      any_whitespace,
      set {
        group.Term,
        group.Factor,

        group.AnchoredExpression,

        capture = true,
      },
    },

    -- The singular items that can be used anywhere as one "unit"
    AnchoredExpression = set {
      group.FuncCall,
      group.ObjectBracketAccess,
      group.ObjectDotAccess,
      group.DictionaryLiteral,
      group.ListLiteral,
      group.StringLiteral,

      group.Number,
      group.Boolean,
      group.Variable,

      group.ParenthedExpression,
    },

    ParenthedExpression = seq {
      any_whitespace,
      "(",
      any_whitespace,
      group.Expression,
      any_whitespace,
      ")",
      any_whitespace,
    },

    Add = literal "+",
    Subtract = literal "-",

    Multiply = literal "*",
    Divide = literal "/",

    Term = seq {
      set { group.Factor, group.AnchoredExpression },
      set { group.Add, group.Subtract },
      group.Expression,

      capture = true,
      linespace = true,
    },

    Factor = seq {
      group.AnchoredExpression,
      set { group.Multiply, group.Divide },
      set { group.Factor, group.AnchoredExpression },

      capture = true,
      linespace = true,
    },

    Unary = set { "+", "-" },

    Number = seq {
      optional { group.Unary },
      set {
        -- Hexadecimal
        seq {
          set { "0x", "0X" },
          some {
            set { digit, range { "a", "f" }, range { "A", "F" } },
          },
        },

        -- Float
        seq {
          some { digit },
          ".",
          some { digit },
        },

        some { digit },
      },

      capture = true,
    },

    Boolean = set {
      "true",
      "v:true",
      "false",
      "v:false",
      capture = true,
    },

    StringLiteral = set {
      --  Courtesy of Lua wiki. Thanks!
      p.literal "'" * ((1 - p.S "'\r\n\f\\") + (p.literal "\\" * 1)) ^ 0 * "'",
      p.literal '"' * ((1 - p.S '"\r\n\f\\') + (p.literal "\\" * 1)) ^ 0 * '"',

      capture = true,
    },

    ObjectBracketAccess = seq {
      set { group.ListLiteral, group.DictionaryLiteral, group.ParenthedExpression },
      "[",
      group.Expression,
      "]",

      linespace = true,
    },

    ObjectDotAccess = seq {
      set { group.DictionaryLiteral, group.ParenthedExpression },
      ".",
      group.VariableIdentifier,

      linespace = true,
    },

    ListLiteral = seq {
      "[",
      list_of { group.Expression, separator = "," },
      "]",

      linespace = true,
    },

    DictionaryKeyExpression = seq {
      "[",
      group.Expression,
      "]",
    },

    DictionaryKey = set {
      group.VariableIdentifier,
      -- TODO: Make the [] expression syntax
      group.DictionaryKeyExpression,
    },

    DictionaryValue = group.Expression,

    DictionaryItem = seq {
      group.DictionaryKey,
      ":",
      group.DictionaryValue,

      linespace = true,
      capture = true,
    },

    DictionaryLiteral = seq {
      "{",
      list_of {
        group.DictionaryItem,

        separator = ",",
        linespace = true,
      },
      "}",
    },

    -- {{{ Function Calls
    FuncCall = seq {
      group.FuncName,
      "(",
      group.FuncCallArgList,
      ")",

      linespace = true,
    },

    FuncName = set {
      group.VariableIdentifier,

      capture = true,
    },

    FuncCallArg = group.Expression,

    FuncCallArgList = one_or_no {
      list_of {
        group.FuncCallArg,

        separator = ",",
        linespace = true,
      },
    },
    -- }}}
  }

  return g.grammar(result)
end

return {
  make_grammar = make_grammar,
}
