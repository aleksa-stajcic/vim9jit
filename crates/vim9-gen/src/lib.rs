use lexer::new_lexer;
use parser;
use parser::new_parser;
use parser::ArrayLiteral;
use parser::AssignStatement;
use parser::AugroupCommand;
use parser::AutocmdCommand;
use parser::Body;
use parser::CallCommand;
use parser::CallExpression;
use parser::DeclCommand;
use parser::DefCommand;
use parser::DictAccess;
use parser::DictLiteral;
use parser::EchoCommand;
use parser::ElseCommand;
use parser::ElseIfCommand;
use parser::ExCommand;
use parser::Expandable;
use parser::Expression;
use parser::GroupedExpression;
use parser::Heredoc;
use parser::Identifier;
use parser::IfCommand;
use parser::IndexExpression;
use parser::IndexType;
use parser::InfixExpression;
use parser::InnerType;
use parser::Lambda;
use parser::Literal;
use parser::MethodCall;
use parser::PrefixExpression;
use parser::RawIdentifier;
use parser::Register;
use parser::ReturnCommand;
use parser::ScopedIdentifier;
use parser::SharedCommand;
use parser::Signature;
use parser::StatementCommand;
use parser::Type;
use parser::UserCommand;
use parser::VarCommand;
use parser::Vim9ScriptCommand;
use parser::VimBoolean;
use parser::VimKey;
use parser::VimNumber;
use parser::VimOption;
use parser::VimScope;
use parser::VimString;

pub mod call_expr;
mod test_harness;

pub struct State {
    pub augroup: Option<Literal>,
    pub is_test: bool,

    pub command_depth: i32,
    pub method_depth: i32,

    // TODO: We could modify the state as we are generating code.
    //  As we generate the code and notice certain identifiers are certain
    //  types, we can use that to do *some* optimizations
    pub scopes: Vec<VimscriptScope>,
}

pub struct VimscriptScope {}

pub trait Generate {
    fn gen(&self, state: &mut State) -> String;
}

impl Generate for ExCommand {
    fn gen(&self, state: &mut State) -> String {
        match self {
            ExCommand::Vim9Script(cmd) => cmd.gen(state),
            ExCommand::Var(cmd) => cmd.gen(state),
            ExCommand::Echo(cmd) => cmd.gen(state),
            ExCommand::Statement(cmd) => cmd.gen(state),
            ExCommand::Return(cmd) => cmd.gen(state),
            ExCommand::Def(cmd) => cmd.gen(state),
            ExCommand::If(cmd) => cmd.gen(state),
            ExCommand::Augroup(cmd) => cmd.gen(state),
            ExCommand::Autocmd(cmd) => cmd.gen(state),
            ExCommand::Call(cmd) => cmd.gen(state),
            ExCommand::Decl(cmd) => cmd.gen(state),
            ExCommand::Comment(token) => format!("\n-- {}", token.text),
            ExCommand::NoOp(token) => format!("\n-- {:?}", token),
            ExCommand::Heredoc(heredoc) => heredoc.gen(state),
            ExCommand::UserCommand(usercmd) => usercmd.gen(state),
            ExCommand::Eval(eval) => format!("{};", eval.expr.gen(state)),
            ExCommand::SharedCommand(shared) => shared.gen(state),
            _ => todo!("Have not yet handled: {:?}", self),
        }
    }
}

impl Generate for SharedCommand {
    fn gen(&self, _: &mut State) -> String {
        format!("vim.cmd [===[{}]===]", self.contents)
    }
}

fn make_user_command_arg(state: &State) -> String {
    format!("__vim9_arg_{}", state.command_depth)
}

impl Generate for UserCommand {
    fn gen(&self, state: &mut State) -> String {
        // TODO:
        // TODO: have not yet handled all the things here
        state.command_depth += 1;

        let result = format!(
            r#"
            vim.api.nvim_create_user_command(
                "{}",
                function({})
                    {}
                end,
                {{
                     nargs = '{}',
                     bang = {},
                }}
            )"#,
            self.name,
            make_user_command_arg(state),
            self.command.gen(state),
            self.command_nargs.clone().unwrap_or("0".to_string()),
            self.command_bang,
        );

        state.command_depth -= 1;

        result
    }
}

impl Generate for DeclCommand {
    fn gen(&self, state: &mut State) -> String {
        // TODO: default value should not be nil for everything here
        // i think?
        format!("local {} = {}", self.name.gen(state), "nil")
    }
}

impl Generate for Heredoc {
    fn gen(&self, state: &mut State) -> String {
        // this works for non-trim and non-eval heredocs perfect
        let inner = self
            .contents
            .iter()
            .map(|line| format!("[==[{}]==]", line))
            .collect::<Vec<String>>()
            .join(", ");

        let mut list = format!("{{ {} }}", inner);
        if self.trim {
            list = format!("require('vim9script').heredoc.trim({})", list)
        }

        format!("local {} = {}", self.name.gen(state), list)
    }
}

impl Generate for CallCommand {
    fn gen(&self, state: &mut State) -> String {
        let call: CallExpression = self.into();
        format!("{};", call.gen(state))
    }
}

impl Generate for AugroupCommand {
    fn gen(&self, state: &mut State) -> String {
        state.augroup = Some(self.augroup_name.clone());

        let group = self.augroup_name.token.text.clone();
        let result = format!(
            r#"
    vim.api.nvim_create_augroup("{}", {{ clear = false }})

    {}
"#,
            group,
            self.body.gen(state)
        );

        state.augroup = None;

        result
    }
}

impl Generate for AutocmdCommand {
    fn gen(&self, state: &mut State) -> String {
        let group = match &state.augroup {
            Some(group) => format!("group = '{}',", group.token.text),
            None => "".to_string(),
        };

        let event_list = self
            .events
            .iter()
            .map(|e| format!("'{}'", e.token.text))
            .collect::<Vec<String>>()
            .join(", ");

        let callback = match &self.block {
            parser::AutocmdBlock::Command(cmd) => {
                format!("function()\n{}\nend", cmd.gen(state))
            }
            parser::AutocmdBlock::Block(block) => {
                format!("function()\n{}\nend", block.body.gen(state))
            }
        };

        format!(
            r#"
vim.api.nvim_create_autocmd({{ {} }}, {{
    {}
    callback = {},
}})
"#,
            event_list, group, callback
        )
    }
}

impl Generate for ReturnCommand {
    fn gen(&self, state: &mut State) -> String {
        match &self.expr {
            Some(expr) => format!("return {}", expr.gen(state)),
            None => "return".to_string(),
        }
    }
}

impl Generate for DefCommand {
    fn gen(&self, state: &mut State) -> String {
        let name = self.name.gen(state);

        if state.is_test && name.starts_with("Test") {
            format!(
                r#"
                    it("{}", function()
                       -- Set errors to empty
                       vim.v.errors = {{}}

                       -- Actual test
                       {}

                       -- Assert that errors is still empty
                       assert.are.same({{}}, vim.v.errors)
                    end)
                "#,
                name,
                self.body.gen(state)
            )
        } else {
            let (signature, default_statements) =
                gen_signature(state, &self.args);

            // TODO: If this command follows certain patterns,
            // we will also need to define a vimscript function,
            // so that this function is available.
            //
            // this could be something just like:
            // function <NAME>(...)
            //   return luaeval('...', ...)
            // endfunction
            //
            // but we haven't done this part yet.
            // This is a "must-have" aspect of what we're doing.
            format!(
                r#"
                    local {} = function({})
                        {}
                        {}
                    end
                "#,
                name,
                signature,
                default_statements,
                self.body.gen(state)
            )
        }
    }
}

fn gen_signature(state: &mut State, args: &Signature) -> (String, String) {
    (
        args.params
            .iter()
            .map(|p| p.name.gen(state))
            .collect::<Vec<String>>()
            .join(", "),
        args.params
            .iter()
            .filter(|p| p.default_val.is_some())
            .map(|p| {
                let name = p.name.gen(state);
                let default_val = p.default_val.clone().unwrap();
                let default_val = default_val.gen(state);

                format!("{name} = vim.F.if_nil({name}, {default_val}, {name})")
            })
            .collect::<Vec<String>>()
            .join("\n"),
    )
}

impl Generate for StatementCommand {
    fn gen(&self, state: &mut State) -> String {
        match self {
            StatementCommand::Assign(assign) => assign.gen(state),
        }
    }
}

impl Generate for AssignStatement {
    fn gen(&self, state: &mut State) -> String {
        format!("{} = {}", self.left.gen(state), self.right.gen(state))
    }
}

impl Generate for IfCommand {
    fn gen(&self, state: &mut State) -> String {
        let else_result = match &self.else_command {
            Some(e) => e.gen(state),
            None => "".to_string(),
        };

        format!(
            r#"
if {} then
  {}
{}
{}
end"#,
            self.condition.gen(state),
            self.body.gen(state),
            self.elseifs
                .iter()
                .map(|e| e.gen(state))
                .collect::<Vec<String>>()
                .join("\n"),
            else_result,
        )
        .trim()
        .to_string()
    }
}

impl Generate for ElseIfCommand {
    fn gen(&self, state: &mut State) -> String {
        format!(
            r#"
        elseif {} then
            {}
        "#,
            self.condition.gen(state),
            self.body.gen(state)
        )
    }
}

impl Generate for ElseCommand {
    fn gen(&self, state: &mut State) -> String {
        format!("else\n{}\n", self.body.gen(state))
    }
}

impl Generate for Body {
    fn gen(&self, state: &mut State) -> String {
        self.commands
            .iter()
            .map(|cmd| cmd.gen(state))
            .collect::<Vec<String>>()
            .join("\n")
    }
}

impl Generate for EchoCommand {
    fn gen(&self, state: &mut State) -> String {
        // TODO: Probably should add some function that
        // pretty prints these the way they would get printed in vim
        // or maybe just call vim.cmd [[echo <...>]] ?
        //  Not sure.
        //  Maybe have to expose something to get exactly the same
        //  results
        //
        // format!("vim.api.nvim_echo({}, false, {{}})", chunks)
        format!("print({})", self.expr.gen(state))
    }
}

impl Generate for Vim9ScriptCommand {
    fn gen(&self, _: &mut State) -> String {
        // TODO: Actually connect this
        // format!("require('vim9script').new_script {{ noclear = {} }}", self.noclear)
        format!("-- vim9script")
    }
}

impl Generate for VarCommand {
    fn gen(&self, state: &mut State) -> String {
        match self.ty {
            Some(Type {
                inner: InnerType::Bool,
                ..
            }) => format!(
                "local {} = require('vim9script').convert.decl_bool({})",
                self.name.gen(state),
                self.expr.gen(state)
            ),
            _ => format!(
                "local {} = {}",
                self.name.gen(state),
                self.expr.gen(state)
            ),
        }
    }
}

impl Generate for Identifier {
    fn gen(&self, state: &mut State) -> String {
        match self {
            Identifier::Raw(raw) => raw.gen(state),
            Identifier::Scope(scoped) => scoped.gen(state),
            Identifier::Unpacked(_) => todo!(),
        }
    }
}

impl Generate for ScopedIdentifier {
    fn gen(&self, state: &mut State) -> String {
        let scope = match self.scope {
            VimScope::Global => "vim.g",
            VimScope::VimVar => "vim.v",
            _ => todo!(),
        };

        format!("{}['{}']", scope, self.accessor.gen(state))
    }
}

impl Generate for RawIdentifier {
    fn gen(&self, _: &mut State) -> String {
        self.name.clone()
    }
}

impl Generate for Expression {
    fn gen(&self, state: &mut State) -> String {
        match self {
            Expression::Identifier(identifier) => identifier.gen(state),
            Expression::Number(num) => num.gen(state),
            Expression::String(str) => str.gen(state),
            Expression::Boolean(bool) => bool.gen(state),
            Expression::Grouped(grouped) => grouped.gen(state),
            Expression::Call(call) => call.gen(state),
            Expression::Prefix(prefix) => prefix.gen(state),
            Expression::Infix(infix) => infix.gen(state),
            Expression::Array(array) => array.gen(state),
            Expression::Dict(dict) => dict.gen(state),
            Expression::VimOption(opt) => opt.gen(state),
            Expression::Empty => "".to_string(),
            Expression::Index(index) => index.gen(state),
            Expression::DictAccess(access) => access.gen(state),
            Expression::Register(reg) => reg.gen(state),
            Expression::Lambda(lambda) => lambda.gen(state),
            Expression::Expandable(expandable) => expandable.gen(state),
            Expression::MethodCall(method) => method.gen(state),

            // Some expressions can only be triggered from containing expressions
            Expression::Slice(_) => unreachable!("cannot gen slice by itself"),
        }
    }
}

impl Generate for MethodCall {
    fn gen(&self, state: &mut State) -> String {
        call_expr::generate_method(self, state)
    }
}

impl Generate for DictAccess {
    fn gen(&self, state: &mut State) -> String {
        format!("{}['{}']", self.container.gen(state), self.index.gen(state))
    }
}

impl Generate for Expandable {
    fn gen(&self, state: &mut State) -> String {
        match &self.ident {
            Identifier::Raw(RawIdentifier { name }) if name == "q-args" => {
                format!("{}.{}", make_user_command_arg(state), "args")
            }
            Identifier::Raw(RawIdentifier { name }) if name == "q-bang" => {
                format!(
                    "({}.{} and '!' or '')",
                    make_user_command_arg(state),
                    "bang"
                )
            }
            _ => format!("vim.fn.expand('<{}>')", self.ident.gen(state)),
        }
    }
}

impl Generate for Lambda {
    fn gen(&self, state: &mut State) -> String {
        let (signature, default_statements) = gen_signature(state, &self.args);
        format!(
            r#"(function({})
                {}
                {}
                end)"#,
            signature,
            default_statements,
            self.body.gen(state)
        )
    }
}

impl Generate for Register {
    fn gen(&self, _: &mut State) -> String {
        format!("vim.fn.getreg({:?})", self.register)
    }
}

impl Generate for PrefixExpression {
    fn gen(&self, state: &mut State) -> String {
        format!(
            "require('vim9script').prefix['{:?}']({})",
            self.operator,
            self.right.gen(state)
        )
    }
}

impl Generate for IndexExpression {
    fn gen(&self, state: &mut State) -> String {
        // println!("self.index = {:#?}", self.index);

        match self.index.as_ref() {
            IndexType::Item(item) => {
                format!(
                    "require('vim9script').index({}, {})",
                    self.container.gen(state),
                    item.gen(state)
                )
            }
            IndexType::Slice(slice) => {
                format!(
                    "require('vim9script').slice({}, {}, {})",
                    self.container.gen(state),
                    slice
                        .start
                        .as_ref()
                        .map_or("nil".to_string(), |item| item.gen(state)),
                    slice
                        .finish
                        .as_ref()
                        .map_or("nil".to_string(), |item| item.gen(state)),
                )
            }
        }
    }
}

impl Generate for VimOption {
    fn gen(&self, state: &mut State) -> String {
        // TODO: not sure if i need to do something smarter than this
        format!(
            "vim.api.nvim_get_option_value('{}', {{}})",
            self.option.gen(state)
        )
    }
}

impl Generate for Literal {
    fn gen(&self, _: &mut State) -> String {
        self.token.text.clone()
    }
}

impl Generate for VimKey {
    fn gen(&self, _: &mut State) -> String {
        match self {
            VimKey::Literal(literal) => literal.token.text.clone(),
        }
    }
}

impl Generate for DictLiteral {
    fn gen(&self, state: &mut State) -> String {
        format!(
            "{{ {} }}",
            self.elements
                .iter()
                .map(|x| format!(
                    "{} = {}",
                    x.key.gen(state),
                    x.value.gen(state)
                ))
                .collect::<Vec<String>>()
                .join(", ")
        )
    }
}

impl Generate for ArrayLiteral {
    fn gen(&self, state: &mut State) -> String {
        format!(
            "{{ {} }}",
            self.elements
                .iter()
                .map(|x| x.gen(state))
                .collect::<Vec<String>>()
                .join(", ")
        )
    }
}

impl Generate for VimString {
    fn gen(&self, _: &mut State) -> String {
        match self {
            VimString::SingleQuote(s) => format!("'{}'", s),
            VimString::DoubleQuote(s) => format!("\"{}\"", s),
        }
    }
}

impl Generate for CallExpression {
    fn gen(&self, state: &mut State) -> String {
        call_expr::generate(self, state)
    }
}

impl Generate for GroupedExpression {
    fn gen(&self, state: &mut State) -> String {
        format!("({})", self.expr.gen(state))
    }
}

impl Generate for VimBoolean {
    fn gen(&self, _: &mut State) -> String {
        format!("{}", self.value)
    }
}

impl Generate for VimNumber {
    fn gen(&self, _: &mut State) -> String {
        self.value.clone()
    }
}

impl Generate for InfixExpression {
    fn gen(&self, state: &mut State) -> String {
        format!(
            "require('vim9script').ops['{:?}']({}, {})",
            self.operator,
            self.left.gen(state),
            self.right.gen(state)
        )

        // match self.operator {
        //     Operator::And => gen_operation(
        //         "AND",
        //         self.left.gen(state),
        //         self.right.gen(state),
        //     ),
        //     Operator::Or => {
        //         gen_operation("OR", self.left.gen(state), self.right.gen(state))
        //     }
        //     // Operator::Plus => todo!(),
        //     // Operator::Minus => todo!(),
        //     // Operator::Bang => todo!(),
        //     // Operator::Modulo => todo!(),
        //     // Operator::Or => todo!(),
        //     // Operator::EqualTo => todo!(),
        //     // Operator::LessThan => todo!(),
        //     // Operator::GreaterThan => todo!(),
        //     // Operator::LessThanOrEqual => todo!(),
        //     // Operator::GreaterThanOrEqual => todo!(),
        //     // Operator::StringConcat => todo!(),
        //     _ => format!(
        //         "({} {} {})",
        //         self.left.gen(state),
        //         self.operator.gen(state),
        //         self.right.gen(state)
        //     ),
        // }
    }
}

pub fn eval(program: parser::Program, is_test: bool) -> String {
    let mut state = State {
        augroup: None,
        command_depth: 0,
        method_depth: 0,
        scopes: vec![],
        is_test,
    };

    let mut output = String::new();
    if is_test {
        output += "describe(\"filename\", function()\n"
    }

    for command in program.commands.iter() {
        output += &command.gen(&mut state);
        output += "\n";
    }

    if is_test {
        output += "end)"
    }

    output
}

pub fn generate(contents: &str, is_test: bool) -> String {
    let lexer = new_lexer(contents);
    let mut parser = new_parser(lexer);
    let program = parser.parse_program();

    let result = eval(program, is_test);
    println!("{}", result);

    let config = stylua_lib::Config::new()
        .with_indent_type(stylua_lib::IndentType::Spaces)
        .with_indent_width(2)
        .with_column_width(120);

    stylua_lib::format_code(
        &result,
        config,
        None,
        stylua_lib::OutputVerification::None,
    )
    .unwrap()
}

#[cfg(test)]
mod test {
    use std::fs::File;
    use std::io::Write;

    use pretty_assertions::assert_eq;

    use super::*;
    use crate::test_harness::exec_busted;
    use crate::test_harness::exec_lua;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(generate(contents, false));
                });
            }
        };
    }

    macro_rules! busted {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let vim_contents = include_str!($path);
                let lua_contents = generate(vim_contents, true);

                let filepath = concat!(
                    env!("CARGO_MANIFEST_DIR"),
                    "/testdata/output/",
                    stringify!($name),
                    ".lua"
                );
                let mut file = File::create(filepath).unwrap();
                write!(file, "{}", lua_contents).unwrap();

                exec_busted(filepath).unwrap();
            }
        };
    }

    busted!(busted_simple_assign, "../testdata/busted/simple_assign.vim");
    busted!(busted_operations, "../testdata/busted/operations.vim");
    busted!(busted_assign, "../testdata/busted/assign.vim");
    busted!(busted_heredoc, "../testdata/busted/heredoc.vim");
    busted!(busted_indexing, "../testdata/busted/indexing.vim");
    busted!(busted_multiline, "../testdata/busted/multiline.vim");
    busted!(busted_function, "../testdata/busted/function.vim");
    busted!(busted_methods, "../testdata/busted/methods.vim");
    busted!(busted_megamethods, "../testdata/busted/megamethods.vim");
    busted!(busted_shared, "../testdata/busted/shared.vim");
    // busted!(busted_vimvars, "../testdata/busted/vimvars.vim");

    snapshot!(test_expr, "../testdata/snapshots/expr.vim");
    snapshot!(test_if, "../testdata/snapshots/if.vim");
    snapshot!(test_assign, "../testdata/snapshots/assign.vim");
    snapshot!(test_call, "../testdata/snapshots/call.vim");
    snapshot!(test_autocmd, "../testdata/snapshots/autocmd.vim");
    snapshot!(test_cfilter, "../testdata/snapshots/cfilter.vim");
    // snapshot!(test_matchparen, "../../shared/snapshots/matchparen.vim");

    #[test]
    fn test_simple_def() {
        let contents = r#"
vim9script

def MyCoolFunc(): number
  return 5
enddef

var x = MyCoolFunc() + 1
"#;

        let generated = generate(contents, false);
        let eval = exec_lua(&generated, "x").unwrap();
        assert_eq!(eval, 6.into());
    }

    #[test]
    fn test_builtin_func() {
        let contents = r#"
vim9script

var x = len("hello")
"#;

        let generated = generate(contents, false);
        let eval = exec_lua(&generated, "x").unwrap();
        assert_eq!(eval, 5.into());
    }

    #[test]
    fn test_augroup_1() {
        let contents = r#"
vim9script

augroup matchparen
  # Replace all matchparen autocommands
  autocmd! CursorMoved,CursorMovedI,WinEnter * {
      echo "Block"
    }

  autocmd WinLeave * echo "Command"
augroup END

var x = len(nvim_get_autocmds({group: "matchparen"}))

"#;

        let generated = generate(contents, false);
        let eval = exec_lua(&generated, "x").unwrap();
        assert_eq!(eval, 4.into());
    }
}
