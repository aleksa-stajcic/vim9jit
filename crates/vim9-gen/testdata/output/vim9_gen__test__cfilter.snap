---
source: crates/vim9-gen/src/lib.rs
assertion_line: 800
expression: "generate(contents, false)"
---
-- vim9script

-- # cfilter.vim: Plugin to filter entries from a quickfix/location list

-- # Last Change: Jun 30, 2022

-- # Maintainer: Yegappan Lakshmanan (yegappan AT yahoo DOT com)

-- # Version: 2.0

-- #

-- # Commands to filter the quickfix list:

-- #   :Cfilter[!] /{pat}/

-- #       Create a new quickfix list from entries matching {pat} in the current

-- #       quickfix list. Both the file name and the text of the entries are

-- #       matched against {pat}. If ! is supplied, then entries not matching

-- #       {pat} are used. The pattern can be optionally enclosed using one of

-- #       the following characters: ', ", /. If the pattern is empty, then the

-- #       last used search pattern is used.

-- #   :Lfilter[!] /{pat}/

-- #       Same as :Cfilter but operates on the current location list.

-- #

-- Token(EndOfLine, "\n", (18,0)->(18,0))

local Qf_filter = function(qf, searchpat, bang)
  local Xgetlist = nil
  local Xsetlist = nil
  local cmd = nil
  local firstchar = nil
  local lastchar = nil
  local pat = nil
  local title = nil
  local Cond = nil
  local items = nil

  -- Token(EndOfLine, "\n", (29,0)->(29,0))
  if qf then
    Xgetlist = function(...)
      return vim.fn["getqflist"](...)
    end
    Xsetlist = function(...)
      return vim.fn["setqflist"](...)
    end
    cmd = require("vim9script").ops["StringConcat"](":Cfilter", bang)
  else
    Xgetlist = function(...)
      local copied = vim.deepcopy({ 0 })
      for _, val in ipairs({ ... }) do
        table.insert(copied, val)
      end
      return vim.fn["function"](unpack(copied))
    end

    Xsetlist = function(...)
      local copied = vim.deepcopy({ 0 })
      for _, val in ipairs({ ... }) do
        table.insert(copied, val)
      end
      return vim.fn["function"](unpack(copied))
    end

    cmd = require("vim9script").ops["StringConcat"](":Lfilter", bang)
  end

  -- Token(EndOfLine, "\n", (39,0)->(39,0))
  firstchar = require("vim9script").index(searchpat, 0)
  lastchar = require("vim9script").slice(searchpat, require("vim9script").prefix["Minus"](1), nil)
  if
    require("vim9script").ops["And"](
      require("vim9script").ops["EqualTo"](firstchar, lastchar),
      (
        require("vim9script").ops["Or"](
          require("vim9script").ops["Or"](
            require("vim9script").ops["EqualTo"](firstchar, "/"),
            require("vim9script").ops["EqualTo"](firstchar, '"')
          ),
          require("vim9script").ops["EqualTo"](firstchar, "'")
        )
      )
    )
  then
    pat = require("vim9script").slice(searchpat, 1, require("vim9script").prefix["Minus"](2))
    if require("vim9script").ops["EqualTo"](pat, "") then
      -- # Use the last search pattern
      pat = vim.fn.getreg("/")
    end
  else
    pat = searchpat
  end

  -- Token(EndOfLine, "\n", (52,0)->(52,0))
  if require("vim9script").ops["EqualTo"](pat, "") then
    return
  end

  -- Token(EndOfLine, "\n", (56,0)->(56,0))
  if require("vim9script").ops["EqualTo"](bang, "!") then
    Cond = function(_, val)
      return require("vim9script").ops["And"](
        require("vim9script").ops["NotRegexpMatches"](val["text"], pat),
        require("vim9script").ops["NotRegexpMatches"](vim.fn["bufname"](val["bufnr"]), pat)
      )
    end
  else
    Cond = function(_, val)
      return require("vim9script").ops["Or"](
        require("vim9script").ops["RegexpMatches"](val["text"], pat),
        require("vim9script").ops["RegexpMatches"](vim.fn["bufname"](val["bufnr"]), pat)
      )
    end
  end

  -- Token(EndOfLine, "\n", (62,0)->(62,0))
  items = require("vim9script").fn_mut("filter", { Xgetlist(), Cond }, { replace = 0 })
  title = require("vim9script").ops["StringConcat"](
    require("vim9script").ops["StringConcat"](require("vim9script").ops["StringConcat"](cmd, " /"), pat),
    "/"
  )
  Xsetlist({}, " ", { title = title, items = items })
end

-- Token(EndOfLine, "\n", (67,0)->(67,0))

vim.api.nvim_create_user_command("Cfilter", function(__vim9_arg_1)
  Qf_filter(true, __vim9_arg_1.args, vim.fn.expand("<q-bang>"))
end, {
  nargs = "+",
})

vim.api.nvim_create_user_command("Lfilter", function(__vim9_arg_1)
  Qf_filter(false, __vim9_arg_1.args, vim.fn.expand("<q-bang>"))
end, {
  nargs = "+",
})

-- Token(EndOfLine, "\n", (70,0)->(70,0))

-- # vim: shiftwidth=2 sts=2 expandtab

